### DISCUSSION TITLE
Retrieve image from Supabase storage via URL

### DISCUSSION THREAD

**Discussion by rgroothuis** [supabase/supabase#13742] [Questions] [answered]
I've images stored in a private Supabase bucket. I know the file names in the buckets and I want to download the images with minimum API requests. So instead of using the "createSignedUrls" for each image download I want to use something like: 

https://xxxxxxxx.supabase.co/object/auth/ ???? /<bucket name>/images.jpg
or
https://xxxxxxxx.supabase.co/storage/ ???? /<bucket name>/images.jpg

Of course I've an API key which I can pass in the header to the Supabase service. The user has authenticated against the Supabase service. 

Any suggestion on the URL should look like? I've searched the internet for a couple of days but couldn't find the solution. Thanks.

**Comment by GaryAustin1** [comment]
Your choices are to use api download or getsignedUrl for a private bucket.  There is no direct URL access without the signedURL.    You could use a uuid coded name in a public bucket to get reasonable security.

**Reply by rgroothuis** [reply]
Thanks for the feedback Gary. How does the api download looks like? Do you have an example?

Or is the api the same thing as the Supabase client would do? The same API calls?

**Reply by GaryAustin1** [reply]
Yes I meant use the client .download or yes you could use the REST interface,  but you need a signed in user JWT in the authorization header.  Mainly saying for private bucket you have to make an API call.

**Reply by rgroothuis** [reply]
I think this is what I’m looking for. But how does this REST JET api call looks like. That is the problem. I cannot find the doc/info.

and I’ve the key of course for the authorization header.

**Reply by GaryAustin1** [reply]
https://supabase.github.io/storage-api/#/

**Reply by rgroothuis** [reply]
Yes, found that, tried that without success. 

I get as response: 

{
    "message": "No API key found in request",
    "hint": "No `apikey` request header or url param was found."
}

This is my postman trial:

<img width="1037" alt="Screenshot 2023-04-15 at 15 20 09" src="https://user-images.githubusercontent.com/14161204/232226433-cee990f5-10d3-4643-89bc-ee71b6ff3eb7.png">

What am I doing wrong?

**Reply by GaryAustin1** [reply]
Seems you also need to set the apikey header to anon key(jwt).

**Reply by rgroothuis** [reply]
Where can I find the value of the apikey? I tried by using the same value as the authorisation key, found at the project settings:

 
<img width="973" alt="Screenshot 2023-04-15 at 22 58 47" src="https://user-images.githubusercontent.com/14161204/232252914-11463fed-e04b-4eca-ae39-997d386b3237.png">

**Reply by GaryAustin1** [reply]
That is the correct value.  You don't use 'bearer' as part of the header for that. 

`apikey: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2.............`

**Reply by rgroothuis** [reply]
Oke, use the same value for the Bearer and APIKEY and came one step further, see screenshot below. 

Now I get the error message: 

{
    "message": "no Route matched with those values"
}

The URL use should give me all the buckets, but it doesn't.

<img width="1362" alt="Screenshot 2023-04-15 at 23 19 48" src="https://user-images.githubusercontent.com/14161204/232253685-c2480b9e-d1be-4179-894e-eadc68e42624.png">

**Reply by GaryAustin1** [reply]
I did not look at your URL earlier....
`https://ugcxxxxxxxxxxxxxxxxxxxxxxq.supabase.co/storage/v1/object/test/public/tesddddt5fffdfffdd49.jpg`
Is an example of a file path.  I assume bucket would go right after the `v1` in that.
I don't use the REST API, but looked at a supabase-js call.

You will also need to meet bucket RLS policy.

**Reply by rgroothuis** [reply]
I'm stuck, I cannot find the correct URL and/or RLS policy. The APIKEY missing error seems to be gone, that is one step forward. But the URL's I'm trying is always refused. See screenshot. 

<img width="1318" alt="Screenshot 2023-04-16 at 12 44 20" src="https://user-images.githubusercontent.com/14161204/232301094-05feca2d-d928-40ff-be2b-6dc8ea87e582.png">

**Reply by GaryAustin1** [reply]
The link I gave you was for a file where you provide the bucket and pathname.  If you are trying to get bucket I mentioned you have to change out object for bucket. Also mine passes in the apikey in a header not as a parameter.

**Reply by rgroothuis** [reply]
Thanks for your patience and getting back to me with suggestions. I'm still being challenged by this problem and have not found the solution yet. 

What I've done so far (with the apikey in the header): 

1) https://xxxxxxxxxxx.supabase.co/storage/v1/object/public/test/mappicture.jpg
this is working fine as the test bucket is a public bucket and I can fetch this file with this url. 

The question is what to change to fetch the files in a private bucket? 

2) the URL: https://xxxxxxxxxxxxx.supabase.co/storage/v1/object/perfectperformance/mappicture.jpg
is giving me the following error response: 

{
    "statusCode": "404",
    "error": "Not found",
    "message": "The resource was not found"
}

The private bucket name is "perfectperformance" and the image file name is: "mappicture.jpg"

In the React Native app the user is authenticated against Supabase. I'm (for now) using Postman to test the URL. Maybe via Postman the user is not authenticated, not sure how to configure that in Postman. 

For testing purposes I changed the perfectperformance bucket from private to public and then the following URL was working: 

https://xxxxxxxxxxxx.supabase.co/storage/v1/object/public/perfectperformance/mappicture.jpg

**Reply by GaryAustin1** [reply]
So with a private bucket you can't just use a URL.  
You have to first have select RLS policy on storage.objects to allow SELECT access to that bucket.
Then you can either download
with GET [...../object/authenticated/{bucketName}/{wildcard}]

Or you can get a signedURL with
POST  [..../object/sign/{bucketName}/{wildcard}]
And then do a get on that signed URL (The url would commonly be for an img src)

**Reply by rgroothuis** [reply]
The bucket policy is what I've set indeed, see screenshot. 

<img width="1130" alt="Screenshot 2023-04-16 at 20 14 59" src="https://user-images.githubusercontent.com/14161204/232333256-f9571a47-08a8-4571-b60a-dee86d2874cc.png">

The problem is maybe that via the Postman application to test the URL I'm not an authenticated user. Is my assumption correct? Or you don't know? I will give this also a try within the React Native application (in the mean time).

**Reply by GaryAustin1** [reply]
You need to replace the anon key in the authorization header with a non-expired user jwt (you got by signing in, or generating with something like jwt.io using the Supabase jwt secret key).
You can do a quick test by changing the 'authenticate' to 'anon' to see if everything else is working with the anon key.

**Reply by rgroothuis** [reply]
Tested within the React Native app where the user is successfully authenticated against the Supabase platform. I'm using the same API key for both the Authorization and the APIKEY in the header. I guess this is not correct? I assume somehow a user token is required, I tested with the access_token received from the supabase.auth.getSession() call but that doesn't work. Other suggestions? 

The whole reason why I want to do this is to save signedURL calls. I'm placing many markers with a picture on a map. By skipping the get signed URLs call I can save 50% of the API calls. Not sure if this is the correct approach? Or should I simply use the signedURL followed by an image download for each marker on the maps? Suggestions?

**Reply by GaryAustin1** [reply]
The access token is what to use in the authorization header.  The apikey header is always anon key.

You can do a bulk load of signedURL's with an array of paths in the body (in same bucket).  Then use the signedURL's.

Not really clear to me where you are downloading.  If you can use the storage-api you just download from a private bucket like a public one using the path name and meet RLS.

This is not the same as just using the image url (signed or public) in another context without the storage api.  signedURL and publicURL do not need any headers they are just image URLs.

**Reply by rgroothuis** [reply, marked as answer]
I've the solution working, the following URL with the header configuration is working. It is downloading the image from the private bucket. 

https://xxxxxxxxxxxx.supabase.co/storage/v1/object/authenticated/perfectperformance/mappicture.jpg

fetch("https://xxxxxxx.supabase.co/storage/v1/object/authenticated/perfectperformance/mappicture.jpg", {
      method : 'GET',
      headers : { 'Content-type' : 'application/json',
                  'apikey' : API_KEY,
                  'Authorization' : 'Bearer ' + user.session.access_token },
    })

Thanks for your support @GaryAustin1 !!

**Reply by ethicnology** [reply]
After struggling myself, here I share a curl command that [you can convert in any language](https://curlconverter.com/)

```sh
curl --request GET \
  --url https://[PROJECT_ID].supabase.co/storage/v1/object/authenticated/[BUCKET]/[ASSET_NAME] \
  --header 'Authorization: Bearer ACCESS_TOKEN' \
  --header 'apikey: ANON_KEY'
```
Placeholder:
- PROJECT_ID: `adfezpyogxtyouzwmmuu`
- BUCKET: `photos`
- ASSET_NAME: `eiffel_tower.jpg`
- ANON_KEY: Supabase `Anon Key`
- ACCESS_TOKEN: `Supabase.instance.client.auth.currentSession.accessToken`

**Reply by bkan36** [reply]
how do you display the image in your html 
I used 

```
            const blob = await data.blob();

            const imageUrl = URL.createObjectURL(blob);

            return imageUrl
```
but it doesn't works
imageUrl is sent to my img src

**Reply by adamyodinsky** [reply]
Why this is not in the docs?

**Reply by ethicnology** [reply]
@adamyodinsky Creating a PR is easy, getting it merge into the supabase repo is long.
[My last PR took 4 months](https://github.com/supabase/supabase/pull/28298) to get merge for two lines

**Comment by Crypto69** [comment]
> I've the solution working, the following URL with the header configuration is working. It is downloading the image from the private bucket.
> 
> https://xxxxxxxxxxxx.supabase.co/storage/v1/object/authenticated/perfectperformance/mappicture.jpg
> 
> fetch("https://xxxxxxx.supabase.co/storage/v1/object/authenticated/perfectperformance/mappicture.jpg", { method : 'GET', headers : { 'Content-type' : 'application/json', 'apikey' : API_KEY, 'Authorization' : 'Bearer ' + user.session.access_token }, })
> 
> Thanks for your support @GaryAustin1 !!



> I've the solution working, the following URL with the header configuration is working. It is downloading the image from the private bucket.
> 
> https://xxxxxxxxxxxx.supabase.co/storage/v1/object/authenticated/perfectperformance/mappicture.jpg
> 
> fetch("https://xxxxxxx.supabase.co/storage/v1/object/authenticated/perfectperformance/mappicture.jpg", { method : 'GET', headers : { 'Content-type' : 'application/json', 'apikey' : API_KEY, 'Authorization' : 'Bearer ' + user.session.access_token }, })
> 
> Thanks for your support @GaryAustin1 !!

How have you found a performance of this? Doing this kind of authorised fetch versus getting a signed url ?
Which solution is more likely to provide cdn cache hits on subsequent calls for the same image?

**Reply by gitnik** [reply]
Would also be interested to hear if the supabase team recommends doing one or the other, especially in cases were you need dozens of assets on a single page

**Reply by GaryAustin1** [reply]
I'm not part of the Supabase team.  Public URL would be the fastest way (even if using hard to guess filenames).  Then a signedURL if used more than once as the method above is going to the API, the database and then s3.  SignedURL went to the API/database.  Then when you use it goes to s3.  That is my understanding of the flow.

**Reply by Crypto69** [reply]
Thanks unfortunately with private photos I don't think either solution is particularly elegant. 

I have an application that users can upload photos. Similar to something like Instagram. Other users should not be allowed to see any of your photos unless they are marked as public.

I setup a private bucket and then create a folder for each user based on their UUID 
i have RLS storage policies to limit access to authorised users and only users that own the folder.

Let's say a user uploads five photos. 
they are now referenced as follows

6k8cc21a-75fb-439a-80abbb-a3c4b0cceb/photos/myphoto1.jpeg
. etc
. etc
. etc
6k8cc21a-75fb-439a-80abbb-a3c4b0cceb/photos/myphoto5.jpeg

if we wants to see their photos now in something like an image carousel.

Because it's a private bucket you can't pass URLs for the images to the carousel.

So now to display all the users images we need to do a few things.

**scenario 1**
1.  api db call to list all the files in the bucket for that user ID
2. Take the array of file names and pass it to createSignedUrls db api
3. pass each signed url link to image carousel which then pulls the image from the S3 bucket

**or**

**scenario 2**
1.  api db call to list all the files in the bucket for that user ID
2. take array of files and call fetch function to download via signed url from S3 to local storage

  https://xxxxxxxxxxxx.supabase.co/storage/v1/object/authenticated/6kpcc21a-89fb-439a-80abbb-a3c4b0cceb/photos/myphoto1.jpeg  and the user Authorization header.

3. pass each downloaded image url that  points to the local downloaded image to the carousel

At this point i think they are similar in performance? Maybe one less API call to the database in scenario two.

What happens if you leave the page and come back 5 min later?

Do either scenario get any CDN hits from the private bucket?

It would be great to hear from anyone who knows what's going on under the hood from a performance perspective.
thanks...

**Comment by amirkhalil22** [comment]
hello, i already have a project that i need to list all documents from a bucket, with a HTTP request

![image](https://github.com/user-attachments/assets/0b5c495d-50ab-4963-a2cb-4da089fa7a55)
this is my bucket named test,

im using this url to get a specific document : https://supabase.nexusteam.net/storage/v1/object/test/cv_med

please need help...
A++
